gitbook-docs/request_response.md

```diff
@@ -0,0 +1,130 @@
+
+- 200 - the request was succesfull
+- 502 - something went wrong carrying out the request on the server side
+- 400 - something is wrong with the client's request
+- 504 - timeout on the server side trying to carry out the request
+- 405 - the server does not support the request
+- 401 - the request has not been applied because it lacks valid authentication credentials
+- 403 - the client does not have permission to make the request
+
+The message can optionally contain a `message`.
+
+The response object may contain other objects depending on the specific request being made. For example, a response to auth request could contain a `login` object.
+
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "COMPLETED",
+  "result": 200,
+  "login": {
+       "token": "....." 
+   }
+}
+```
+
+A server may respond to a request multiple times depending on how it processes the request.
+
+When a server cannot process the request immediately, it will respond with the `state` PENDING:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "PENDING"
+}
+```
+
+When processing is done, but it was not succesfull:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "COMPLETED",
+  "result": 502,
+  "message": "Unable to contact the light"
+}
+```
+
+When processing completed successfully:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "COMPLETED",
+  "result": 200
+}
+```
+
+The state of a request can also be found by sending the following:
+
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "query": true
+}
+```
+
+This will result in a reply like the examples above.
+
+## HTTP
+
+HTTP request use REST api semantics and the reponses are similar to the `response` object used above.
+
+One difference is that the `result` value above is sent as the HTTP response code.
+
+The response when a server succcesfully processes a login request synchronously:
+
+HTTP response code 200
+```json
+{
+  "state": "COMPLETED",
+  "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkZXZpY2UiOiIxMjM0LTQ1NjUzLTM0MzQ1MyIsImlhdCI6MTUzNjg4NDY5MSwiZXhwIjoxNTY4NDQyMjkxfQ.5wypdKin5Q-gsi9aQ8sN1XBAP8bt3tNBT1WiIttm3qM"
+}
+```
+
+When a request is PENDING, an HTTP 202 (Acepted) code will be returned and the body will include an `href` to use to check the status of the request:
+
+HTTP response code 202
+```json
+{
+  "state": "PENDING",
+  "href": "/signalk/v1/api/actions/12567"  
+}
+```
+
+The contents of the response message when checking the status will include the values defined above for the `result` object and may also include extra information related to the request.
+
+For example, the result of a PUT request:
+```json
+{
+   "state": "COMPLETED",
+   "result": 200
+}
+```
```

Comment:
Shouldn't these last two contain `requestId`, so that the envelope is consistent across the board?

Reply:
I'm not sure that `context` is always relevant. For example, I don't think makes sense in the device access request stuff. 

For HTTP, the client never needs the `requestId`, is I have left it out.

Reply:
See my question above - I don't think `context` is part of the request response protocol, but part of the payload.

Reply:
@tkurki that makes sense to me

Reply:
removed